android/KMEA/app/src/main/java/com/tavultesoft/kmea/logic/ResourcesUpdateTool.java

```diff
@@ -0,0 +1,488 @@
+      dialogBuilder.setTitle(currentContext.getString(R.string.keyboard_updates_available));
+      dialogBuilder.setMessage(currentContext.getString(R.string.confirm_update));
+      dialogBuilder.setPositiveButton(currentContext.getString(R.string.label_update), new DialogInterface.OnClickListener() {
+        public void onClick(DialogInterface dialog, int which) {
+          // Update keyboards
+          if (KMManager.hasConnection(currentContext)) {
+            // For each updatable keyboard, one at a time, do the update.
+            // TODO:  May need a reework to better handle large amounts of updates -
+            //        these calls will stack within the Android subsystem and may have issues accordingly.
+            for (Bundle resourceBundle : updatableResources) {
+
+             addOpenUpdate(resourceBundle);
+
+              Intent intent = new Intent(currentContext, KMKeyboardDownloaderActivity.class);
+              intent.putExtras(resourceBundle);
+              currentContext.startActivity(intent);
+            }
+          } else {
+            Toast.makeText(currentContext, "No internet connection", Toast.LENGTH_SHORT).show();
+            checkingUpdates = false;
+          }
+        }
+      });
+
+      dialogBuilder.setNegativeButton(currentContext.getString(R.string.label_later), new DialogInterface.OnClickListener() {
+        public void onClick(DialogInterface dialog, int which) {
+          lastUpdateCheck = Calendar.getInstance();
+          checkingUpdates = false;
+        }
+      });
+
+      AlertDialog dialog = dialogBuilder.create();
+      if (isContextAvailable()) {
+        dialog.setOnCancelListener(new DialogInterface.OnCancelListener() {
+          @Override
+          public void onCancel(DialogInterface dialog) {
+            lastUpdateCheck = Calendar.getInstance();
+            checkingUpdates = false;
+          }
+        });
+        dialog.show();
+      } else {
+        checkingUpdates = false;
+      }
+    }
+  }
+
+  /**
+   * Add open update.
+   * @param anUpdateId the update id
+   * @param aNotificationId the system notification id
+   * @param aBundle the bundle
+   */
+  private void addOpenUpdate(String anUpdateId,Integer aNotificationId, Bundle aBundle)
+  {
+    int _old= openUpdates.size();
+    OngoingUpdate _update = new OngoingUpdate();
+    _update.updateid = anUpdateId;
+    _update.notificationid = aNotificationId;
+    _update.bundle = aBundle;
+    openUpdates.put(anUpdateId, _update);
+    propertyChangeSupport.firePropertyChange("updateCount",_old,openUpdates.size());
+  }
+
+  /**
+   * remove the open update package.
+   * cancels the system notification.
+   * @param anUpdateId the update id
+   */
+  private void removeOpenUpdate(String anUpdateId)
+  {
+    int _old= openUpdates.size();
+    OngoingUpdate _update = openUpdates.remove(anUpdateId);
+
+    if(_update.notificationid!=null) {
```

Comment:
I got an NPE here.
Scenario was a build with "older" sil_euro_latin keyboard and "older" nrc.en.mtnt model.
App came up and showed 2 updates to install. I clicked "Download" for the model when the exception hit.